docs/migration-2.0.md

```diff
@@ -0,0 +1,1178 @@
+
+The `bsonrw.NewBSONDocumentReader` has been renamed to `NewDocumentReader`, which reads from an `io.Reader`, in the `bson` package.
+
+The `NewBSONValueReader` has been removed.
+
+This example creates a `Decoder` that reads from a byte slice.
+
+```go
+// v1
+
+b, _ := bson.Marshal(bson.M{"isOK": true})
+decoder, err := bson.NewDecoder(bsonrw.NewBSONDocumentReader(b))
+```
+
+```go
+// v2
+
+b, _ := bson.Marshal(bson.M{"isOK": true})
+decoder := bson.NewDecoder(bson.NewDocumentReader(bytes.NewReader(b)))
+```
+
+### NewBSONValueWriter
+
+The `bsonrw.NewBSONValueWriter` function has been renamed to `NewDocumentWriter` in the `bson` package.
+
+This example creates an `Encoder` that writes BSON values to a `bytes.Buffer`.
+
+```go
+// v1
+
+buf := new(bytes.Buffer)
+vw, err := bsonrw.NewBSONValueWriter(buf)
+encoder, err := bson.NewEncoder(vw)
+```
+
+```go
+// v2
+
+buf := new(bytes.Buffer)
+vw := bson.NewDocumentWriter(buf)
+encoder := bson.NewEncoder(vw)
+```
+
+## Mgocompat Package
+
+The `bson/mgocompat` has been simplified. Its implementation has been merged into the `bson` package.
+
+`ErrSetZero` has been renamed to `ErrMgoSetZero` in the `bson` package.
+
+`NewRegistryBuilder` function has been simplified to `NewMgoRegistry` in the `bson` package.
+
+Similarly, `NewRespectNilValuesRegistryBuilder` function has been simplified to `NewRespectNilValuesMgoRegistry` in the `bson` package.
+
+## Primitive Package
+
+The `bson/primitive` package has been merged into the `bson` package.
+
+Additionally, the `bson.D` has implemented the `json.Marshaler` and `json.Unmarshaler` interfaces, where it uses a key-value representation in "regular" (i.e. non-Extended) JSON.
+
+The `bson.D.String` and `bson.M.String` methods return a relaxed Extended JSON representation of the document.
+
+```go
+// v2
+
+d := D{{"a", 1}, {"b", 2}}
+fmt.Printf("%s\n", d)
+// Output: {"a":{"$numberInt":"1"},"b":{"$numberInt":"2"}}
+```
+
+## Bson Package
+
+### DefaultRegistry / NewRegistryBuilder
+
+`DefaultRegistry` has been removed. Using `bson.DefaultRegistry` to either access the default registry behavior or to globally modify the default registry, will be impacted by this change and will need to configure their registry using another mechanism.
+
+The `NewRegistryBuilder` function has been removed along with the `bsoncodec.RegistryBuilder` struct as mentioned above.
+
+### Decoder
+
+#### NewDecoder
+
+The signature of `NewDecoder` has been updated without an error being returned.
+
+#### NewDecoderWithContext
+
+`NewDecoderWithContext` has been removed in favor of using the `SetRegistry` method to set a registry.
+
+Correspondingly, the following methods have been removed:
+
+- `UnmarshalWithRegistry`
+- `UnmarshalWithContext`
+- `UnmarshalValueWithRegistry`
+- `UnmarshalExtJSONWithRegistry`
+- `UnmarshalExtJSONWithContext`
+
+For example, a boolean type can be stored in the database as a BSON boolean, int32, or int64. Given a registry:
```

Comment:
```suggestion
For example, a boolean type can be stored in the database as a BSON `Boolean`, `Int32`, or `Int64`. Given a registry:
```

Reply:
Go native types are not capitalized.

Reply:
I regard "boolean", "in32", "int64", or "true" are general concepts, that don't need to be denoted.

Reply:
@prestonvasquez isn't this specifically talking about the BSON types rather than the native types? This suggestion was based on the code sample below which has `bsontype.Boolean`, `bsontype.Int32`, and `bsontype.Int64`

Reply:
Ok, I'll defer to your judgement here. I noticed that it was capitalized in the code sample, which is why I suggested it, but if it doesn't make sense, let's keep it as-is. Feel free to resolve this one :)